apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_subscriber.erl

```diff
@@ -0,0 +1,473 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+%% @doc State machine for a single subscription of a shared subscription agent.
+%% Implements Shared subscriber described in
+%% https://github.com/emqx/eip/blob/main/active/0028-durable-shared-subscriptions.md
+
+%% Referred as `ssubscriber` outside of this module
+-module(emqx_ds_shared_sub_subscriber).
+
+-include_lib("emqx/include/logger.hrl").
+-include("emqx_ds_shared_sub_config.hrl").
+-include("emqx_ds_shared_sub_proto.hrl").
+-include("emqx_ds_shared_sub_format.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+-export([
+    new/1,
+
+    %% Self-initiated/leader messages
+    on_info/2,
+
+    %% Callbacks from the enclosing session
+    on_stream_progress/2,
+    on_disconnect/2,
+    on_unsubscribe/1
+]).
+
+-export_type([
+    t/0,
+    options/0
+]).
+
+-type options() :: #{
+    session_id := emqx_persistent_session_ds:id(),
+    id := emqx_ds_shared_sub_proto:ssubscriber_id(),
+    share_topic_filter := emqx_persistent_session_ds:share_topic_filter(),
+    send_after := fun((non_neg_integer(), term()) -> reference())
+}.
+
+-type progress() :: emqx_persistent_session_ds_shared_subs:progress().
+
+%% Events that SSsubscriber propagates to the Agent which in turn
+%% updates the session.
+
+-type to_agent_events() ::
+    #{
+        type := lease,
+        stream := emqx_ds:stream(),
+        progress := progress()
+    }
+    | #{
+        type := revoke,
+        stream := emqx_ds:stream()
+    }.
+
+%% SSubscriber states
+
+-define(connecting, connecting).
+-define(connected, connected).
+-define(unsubscribing, unsubscribing).
+
+-type status() :: ?connecting | ?connected | ?unsubscribing.
+
+-define(is_connecting(St), (map_get(status, St) =:= ?connecting)).
+-define(is_connected(St), (map_get(status, St) =:= ?connected)).
+-define(is_unsubscribing(St), (map_get(status, St) =:= ?unsubscribing)).
+
+%% Possible status transitions are:
+%% connecting -> connected -> unsubscribing
+
+%% Individual stream states
+
+%% Stream is granted by the leader, and we are consuming it
+-define(stream_granted, stream_granted).
+%% Stream is being revoked by the leader, and we are waiting for the session
+%% to reach a consistent state of stream consumption.
+-define(stream_revoking, stream_revoking).
+
+%% Possible stream transitions are as follows:
+%%
+%% [no stream] -> stream_granted -> stream_revoking -> [no stream]
+
+-type stream_status() :: ?stream_granted | ?stream_revoking.
+
+-type stream_data() :: #{
+    status := stream_status(),
+    progress := progress(),
+    use_finished := boolean()
```

Comment:
Nit. I think a comment describing what it is and what it is for would be welcome here.